docs/specs/user-model.md

```diff
@@ -1,71 +1,127 @@
+
+- `email`
+- `phone_number`
+- `preferred_username`
+
+All of them are nullable.
+
+The user can select their `email`, `phone_number` and `preferred_username` from the identity claims in the settings page.
+
+The admin can do the same thing in the portal.
+
+### Standard attributes - email
+
+When the `email` standard attribute is null, adding or updating identity will copy the `email` claim to the `email` standard attribute.
+
+When the user loses the last ownership of an `email` identity claim, the `email` standard attribute is set to the `email` claim of the oldest identity if present, or null if absent.
+
+### Standard attributes - phone\_number
+
+When the `phone_number` standard attribute is null, adding or updating identity will copy the `phone_number` claim to the `phone_number` standard attribute.
+
+When the user loses the last ownership of an `phone_number` identity claim, the `phone_number` standard attribute is set to the `phone_number` claim of the oldest identity if present, or null if absent.
+
+### Standard attributes - preferred\_username
+
+When the `preferred_username` standard attribute is null, adding or updating identity will copy the `preferred_username` claim to the `preferred_username` standard attribute.
+
+When the user loses the last ownership of an `preferred_username` identity claim, the `preferred_username` standard attribute is set to the `preferred_username` claim of the oldest identity if present, or null if absent.
+
+> TODO: Should we promote the `email` identity claim into `preferred_username` standard attribute?
+
+## Custom Attributes
+
+Custom Attributes is a JSON Object. The developer can update it via Admin API.
+
+Custom Attributes can be any valid JSON Object. There is no limitation on the nesting level. However, a sensible size limit (e.g. 10MiB) is enforced.
+
+The developer can save whatever they want in Custom Attributes.
+
+The user cannot read or write Custom Attributes directly.
+
+> NOTE: The initial implementation is very likely to be a single textarea for editing the whole custom attributes object.
+> Much more fancier features like generated form fields requires us to interpret the JSON schema and generate form fields.
+> That could be a very useful standalone product.
+
+### Custom Attributes validation
+
+The validation on custom attributes can be done with the JSON schema provided by the developer in the configuration.
+
+The JSON schema must be written against [the version 2019-09](https://json-schema.org/specification-links.html#2019-09-formerly-known-as-draft-8).
+
+For example,
+
+```yaml
+custom_attributes:
+  json_schema:
+    type: object
+    properties:
+      email:
+        type: string
+        format: email
+    required: ["email"]
+```
+
+Any update on custom attributes must be valid against the given schema.
+
+> NOTE: In the future, we can add our own JSON schema keyword to annotate the access control of individual fields in the custom attributes.
```

Comment:
As a 2nd thought, I think we should include access control depends on Client ID on this spec.

Because I have a gut feeling that it might complicate the claims computation x custom attributes a lot (e.g. if a custom attribute is not accessible by Client A, but it is still in the claims, what would happen?) We can leave the implementation to later, but I hope our spec can cater it so we don't introduce a big backward incompatibility and redo everything in future...

Reply:
If we provide access control on custom attributes/user claims, I think we'll want to provide access control on APIs for OAuth clients as well.

Usually OAuth scopes are hierarchical, with possibly complex interpretation (e.g. `user.email`, `user.email.read`). We may need to take that into account when designing the access control mechanism.

Reply:
I have added a document about access control.